src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.Match.Tests.cs

```diff
@@ -1269,6 +1269,47 @@ public void Match_CachedPattern_NewTimeoutApplies(RegexOptions options)
             Assert.InRange(sw.Elapsed.TotalSeconds, 0, 10); // arbitrary upper bound that should be well above what's needed with a 1ms timeout
         }
 
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNetCore))]
+        public void NonBacktracking_NoEndAnchorMatchAtTimeoutCheck()
+        {
+            // First figure out how many characters the innermost matching loop of NonBacktracking looks at between
+            // timeout checks. This makes the test more robust, as the value isn't exposed and we'd have to set it as
+            // a constant here.
+            Regex timeoutPattern = new Regex("a*", RegexHelpers.RegexOptionNonBacktracking, TimeSpan.FromTicks(1));
+            // Function for comparing the left hand side to the internal chars matched per timeout check in NonBacktracking
+            int IsCharsPerTimeoutCheck(int candidate, int dummy)
+            {
+                try
+                {
+                    timeoutPattern.Match(new string('a', candidate));
+                    try
+                    {
+                        timeoutPattern.Match(new string('a', candidate + 1));
+                    }
+                    catch (RegexMatchTimeoutException)
+                    {
+                        // Only the higher check timed out, number of chars per timeout check found
+                        return 0;
+                    }
+                    // Neither check timed out, chars per timeout check is higher
+                    return -1;
+                }
+                catch (RegexMatchTimeoutException)
+                {
+                    // Candidate times out, chars per timeout check is lower
+                    return 1;
+                }
+            }
+            int charsPerTimeoutCheck = Array.BinarySearch(Enumerable.Range(0, 16_000).ToArray(), -1, Comparer<int>.Create(IsCharsPerTimeoutCheck));
```

Comment:
I'll simplify the test by hardcoding the `charsPerTimeoutCheck` and including asserts that validate it's indeed the right value before the actual test.

Reply:
I simplified the test even further, now there's just a constant that has to be large enough to trigger timeout checks. I realized it doesn't have to be exact for the test pattern `^a*$`, as treating a timeout check in any point of a string "aaaaaaa...aaab" as an end-of-input would cause a match. I verified that the test indeed does fail if I reintroduce the slicing.